Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp

```diff
@@ -206,82 +206,87 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
+    auto sourceColorSpace = image.colorSpace();
+    auto destinationColorSpace = sourceColorSpace.supportsOutput() ? sourceColorSpace : DestinationColorSpace::SRGB();
     for (auto length : lengths) {
-        RetainPtr context = adoptCF(CGBitmapContextCreate(nullptr, length, length, 8, length * 4, colorSpace.get(), kCGImageAlphaPremultipliedLast));
-        if (!context)
-            return nullptr;
-
-        // Quality is more important than rendering speed for the current use case.
-        CGContextSetInterpolationQuality(context.get(), kCGInterpolationHigh);
-        CGContextDrawImage(context.get(), CGRectMake(0, 0, length, length), image.platformImage().get());
-        RetainPtr newImage = adoptCF(CGBitmapContextCreateImage(context.get()));
-        if (!newImage)
-            return nullptr;
-
-        CGImageDestinationAddImage(destination.get(), newImage.get(), nullptr);
+        if (RefPtr bitmap = ShareableBitmap::createFromImageDraw(image, destinationColorSpace, { (int)length, (int)length }, ShareableBitmap::ShouldShrinkToFit::Yes)) {
+            bitmaps.append(WTFMove(bitmap));
+            continue;
+        }
+        return { };
     }
 
-    if (!CGImageDestinationFinalize(destination.get()))
-        return nullptr;
-
-    return SharedBuffer::create(destinationData.get());
+    return bitmaps;
 }
 
-static RefPtr<SharedBuffer> expandSVGImageToData(SVGImage& svgImage, ASCIILiteral uti, std::span<const unsigned> lengths)
+static Vector<RefPtr<ShareableBitmap>> getBitmapsFromSVGImage(SVGImage& image, std::span<const unsigned> lengths)
 {
-    RetainPtr colorSpace = adoptCF(CGColorSpaceCreateWithName(kCGColorSpaceSRGB));
-    RetainPtr destinationData = adoptCF(CFDataCreateMutable(0, 0));
-    RetainPtr cfUTI = uti.createCFString();
-    RetainPtr destination = adoptCF(CGImageDestinationCreateWithData(destinationData.get(), cfUTI.get(), lengths.size(), nullptr));
+    Vector<RefPtr<ShareableBitmap>> bitmaps;
     for (auto length : lengths) {
-        auto size = FloatSize { float(length), float(length) };
-        RefPtr buffer = ImageBuffer::create(size, RenderingMode::Unaccelerated, RenderingPurpose::Unspecified, 1, DestinationColorSpace::SRGB(), ImageBufferPixelFormat::BGRA8);
-        if (!buffer)
-            return nullptr;
-
-        Ref svgImageContainer = SVGImageForContainer::create(&svgImage, size, 1, { });
-        buffer->context().drawImage(svgImageContainer.get(), FloatPoint::zero());
-
-        RefPtr newImage = ImageBuffer::sinkIntoNativeImage(WTFMove(buffer));
-        if (!newImage)
-            return nullptr;
-
-        CGImageDestinationAddImage(destination.get(), newImage->platformImage().get(), nullptr);
+        IntSize size { (int)length, (int)length };
+        if (RefPtr nativeImage = image.nativeImage(image.size())) {
+            if (RefPtr bitmap = ShareableBitmap::createFromImageDraw(*nativeImage, DestinationColorSpace::SRGB(), size, ShareableBitmap::ShouldShrinkToFit::Yes)) {
```

Comment:
Unless I'm missing something this refactor moves from vector scaling to bitmap scaling. We should probably continue to use `SVGImageForContainer`. Anyone ideas for how to add test coverage for that?

Reply:
I updated the patch to use `SVGImageForContainer` (in new function `nativeImageFromSVGImage`).
Also updated the test to ensure image data is expected.